.github/workflows/publish.yml

```diff
@@ -0,0 +1,121 @@
+name: Publish Docker image
+
+on:
+  push:
+    tags:
+      - 'v*'
+    branches:
+      - main
+
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}
+
+permissions:
+  contents: read
+
+jobs:
+
+  test:
+    name: Test
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    services:
+      postgres:
+        image: postgres
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+        ports:
+          - 5432:5432
+        options: --health-cmd="pg_isready" --health-interval=10s --health-timeout=5s --health-retries=3
+
+    steps:
+      - name: Install packages
+        run: sudo apt-get update && sudo apt-get install --no-install-recommends -y google-chrome-stable curl libvips postgresql-client libpq-dev
+
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Set up Ruby
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: .ruby-version
+          bundler-cache: true
+
+      - name: Run tests and smoke test seed
+        env:
+          RAILS_ENV: test
+          DATABASE_URL: postgres://postgres:postgres@localhost:5432
+        run: |
+          bin/rails db:create
+          bin/rails db:schema:load
+          bin/rails test:all
+          bin/rails db:seed
+
+      - name: Keep screenshots from failed system tests
+        uses: actions/upload-artifact@v4
+        if: failure()
+        with:
+          name: screenshots
+          path: ${{ github.workspace }}/tmp/screenshots
+          if-no-files-found: ignore
+
+  build:
+    name: Build docker image
+    runs-on: ubuntu-latest
+    needs: [test]
+    permissions:
+      packages: write
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v4
+
+      - name: Set up Docker Context for Buildx
+        run: docker context create builders
+        shell: bash
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+        with:
+          endpoint: builders
+
+      - name: Log in to the container registry
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+          flavor: latest=auto
+          tags: |
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{major}}.{{minor}}.{{patch}}
+            type=ref,event=branch
```

Comment:
In my experience, the `latest` tag always points to the latest release. Do you plan for the main branch to always be production-ready?

Reply:
In theory (since we're using trunk-based development), yes, `main` should always be production ready.  If it's not, the checks should fail and the Docker image should not be published.

I would expect `latest` to represent the _most recently published image_, which in our case, would always be the latest commit.

But there probably is some value in providing an alias for self-hosters who want to use the "latest release".  What do you think of using the following tag scheme?

- `latest` - the latest published image (i.e. `main`)
- `stable` - the latest published release
- `[release_semver_tag]` - the semver tag of the release
- `[commit_sha]` - the commit sha of the docker image so self-hoster can lock into a certain version

So something like:

```yml
tags:
  # Explicit commit sha and version tags
  type=sha
  type=semver,pattern={{version}}

  # Aliases
  type=raw,value=latest,enable=${{ github.ref == 'refs/heads/main' }}
  type=raw,value=stable,enable=${{ startsWith(github.ref, 'refs/tags/v') }}
```

Reply:
As long as the tags are semvar compatible, the existing tag scheme is very nice to be able to select an upgrade schedule using auto image pulls. One can have `maybe:latest` which always auto upgrades (and should not be done), while one can do `maybe:2` or `maybe:2.3` and get the patch upgrades.

Reply:
That's how I've seen it done in other projects. It's intuitive, and also works well with renovate/dependabot. But I guess that would impose stricter adherence to semver versioning to be totally useful, so it's more for the core maintainers to decide. It can always be added later :D